### DISCUSSION TITLE
Arduino-esp32 Zigbee example: switch end device

### DISCUSSION THREAD

**Discussion by Antonia-place** [espressif/arduino-esp32#10344] [Q&A]
Hello,
I am new into Zigbee network and I want to create a mini-network using ESP32-C6 as end devices and an Linux machine with Zigbee2Mqtt and Node-Red as coordinator and display.
I found the examples from here, and I used until now the temperature sensor and light lamp: https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Zigbee
Now, I want to add another esp32-c6 acting like an end device switch but I don't know how to do that. I start to modify the temperature sensor example without any results.
This is my work until now: 
[Zigbee_switch_05.zip](https://github.com/user-attachments/files/16995148/Zigbee_switch_05.zip)
In the ESP32 Zigbee example forled I see an similar example but is set as coordinator and this is not good for my network, because I need an switch as end device.
Could you, please, help me to solve this issue regarding to an zigbee switch example end device?
Thank you, 
Mihai

**Comment by VojtechBartoska** [comment]
Hello @Antonia-place,

I am not sure about your question.

BTW. Our implementation of Zigbee library is in progress here: https://github.com/espressif/arduino-esp32/pull/10265

@P-R-O-C-H-Y can provide more details

**Reply by Antonia-place** [reply]
Hello @VojtechBartoska ,
My question is related to the modification of switch example toi act as an end device only, because i want to use it with Zigbee2Mqtt and Node-Red on a Linux Machine.
I tried to do this but I see that all examples uses FreeRTOS and is not clear for me what shlud modify there.

Thank you,
Mihai

**Comment by P-R-O-C-H-Y** [comment]
Hi @Antonia-place, the examples in `libraries/ESP32/examples/Zigbee` are always made in pairs so they can run on 2 ESP32s devices as standalone zigbee network. Thats why on/off light example is set to be an end device and switch as coordinator.

For sure you can edit the [Zigbee_Light_Switch](https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Zigbee/Zigbee_Light_Switch) example to act as end device or router. 

Just to inform you those examples will be removed soon, as there is a Zigbee library pending PR #10265 with its own examples, that may be a lot easier to use. You can take a look there.

About your case right know, I assume to use the on/off light + on/off switch example, extract the Zigbee initialisation part of the end device one and replace it in the switch example.

**Reply by Antonia-place** [reply]
Hello @P-R-O-C-H-Y P_R,
I tested the examples Light_Bulb and Light_Switch, both flashed on ESP32-V3, with succes. Moreover, I connected the Light_bulb to Zigbee2Matt and I was able to send the On/Off state from Node-Red. Now, I want to modify the Light_Switch to act as an end device, to connect it to Zigbee2mqtt, but I can do that because I am not shure how to properly modify the example from coordinator to end-device. In the first topic I attached my modifications for this.
I tryed to change only the ESP_ZB_ZED_CONFIG, ZIGBEE_MODE_ED (from the very begining of the code) and add ED_AGING_TIMEOUT and ED_KEEP_ALIVE in /* Zigbee configuration */ se ction; the code compile with succes, in arduino console i see the respons from the press on-board button; in Zigbee2mqtt I "see" the device but no modification of status button is present.
Could you, please, give me some hints what should I modify to achive the transformation from coordinator to end device, for the Light_Switch Example? I mean explicitly which lines to modify/add and where?

Thank you,
Mihai

**Reply by kcajjones** [reply]
I'm also struggling with this. I'm using Copilot to try and help code but it's not working. An example switch end device (or even better, Router) would be great as I just want to make a few smart plugs with my ESP32-C6 over zigbee controlled via Home Assistant.

**Reply by Dronedale** [reply]
Hi
I'm also trying to use the c6 as an end device switch to integrate it in z2m and use it as a remote control for my ceiling lights with no success.
If someone has some working code for this, especially with color control/ dimming feature that would be nice!

**Reply by Dronedale** [reply]
got it working finally.. simply had been to far away from my coordinator.
here´s my working code, slightly modified example code to act as end device instead of coordinator:

// Copyright 2024 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @brief This example demonstrates Zigbee color dimmer switch.
 *
 * The example demonstrates how to use Zigbee library to control a RGB light bulb.
 * The RGB light bulb is a Zigbee end device, which is controlled by a Zigbee coordinator (Switch).
 * To turn on/off the light, push the button on the switch.
 * To change the color or level of the light, send serial commands to the switch.
 *
 * By setting the switch to allow multiple binding, so it can bind to multiple lights.
 * Also every 30 seconds, all bound lights are printed to the serial console.
 *
 * Proper Zigbee mode must be selected in Tools->Zigbee mode
 * and also the correct partition scheme must be selected in Tools->Partition Scheme.
 *
 * Please check the README.md for instructions and more detailed description.
 *
 * Created by Jan Procházka (https://github.com/P-R-O-C-H-Y/)
 */

#ifndef ZIGBEE_MODE_ED
#error "Zigbee coordinator mode is not selected in Tools->Zigbee mode"
#endif

#include "Zigbee.h"

/* Zigbee color dimmer switch configuration */
#define SWITCH_ENDPOINT_NUMBER 1
uint8_t button = BOOT_PIN;

/* Zigbee switch */
ZigbeeColorDimmerSwitch zbSwitch = ZigbeeColorDimmerSwitch(SWITCH_ENDPOINT_NUMBER);


/********************* Arduino functions **************************/
void setup() {
  //Zigbee.factoryReset();

  Serial.begin(115200);

  //Init button switch
  pinMode(button, INPUT_PULLUP);

  //Optional: set Zigbee device name and model
  zbSwitch.setManufacturerAndModel("Espressif", "ZBLightBulb");

  //Optional to allow multiple light to bind to the switch
  //zbSwitch.allowMultipleBinding(true);

  //Add endpoint to Zigbee Core
  Zigbee.addEndpoint(&zbSwitch);

  //Open network for 180 seconds after boot
  //Zigbee.setRebootOpenNetwork(180);

  //When all EPs are registered, start Zigbee with ZIGBEE_COORDINATOR mode
  if (!Zigbee.begin()) {
    Serial.println("Zigbee failed to start!");
    Serial.println("Rebooting...");
    ESP.restart();
  }

  Serial.println("Connecting to network");
  while (!Zigbee.connected()) {
    Serial.print(".");
    delay(100);
  }
  Serial.println();
}

void loop() {
  // Handle button switch in loop()
  if (digitalRead(button) == LOW) {  // Push button pressed
    // Key debounce handling
    while (digitalRead(button) == LOW) {
      delay(50);
    }
    // Toggle light
    Serial.println("Button pressed");
    zbSwitch.toggle();
  }
  // Handle serial input to control color and level of the light
  if (Serial.available()) {
    String command = Serial.readString();
    Serial.println(command);
    if (command == "on/n") {
      zbSwitch.lightOn();
    } else if (command == "off") {
      zbSwitch.lightOff();
    } else if (command == "toggle") {
      zbSwitch.lightToggle();
    } else if (command == "red") {
      zbSwitch.setLightColor(255, 0, 0);
    } else if (command == "green") {
      zbSwitch.setLightColor(0, 255, 0);
    } else if (command == "blue") {
      zbSwitch.setLightColor(0, 0, 255);
    } else if (command == "white") {
      zbSwitch.setLightColor(255, 255, 255);
    } else if (command == "color") {
      //wait for color value
      Serial.println("Enter red value (0-255):");
      while (!Serial.available()) {
        delay(100);
      }
      int red = Serial.parseInt();
      Serial.println("Enter green value (0-255):");
      while (!Serial.available()) {
        delay(100);
      }
      int green = Serial.parseInt();
      Serial.println("Enter blue value (0-255):");
      while (!Serial.available()) {
        delay(100);
      }
      int blue = Serial.parseInt();
      zbSwitch.setLightColor(red, green, blue);
    } else if (command == "level") {
      //wait for level value
      Serial.println("Enter level value (0-255):");
      while (!Serial.available()) {
        delay(100);
      }
      int level = Serial.parseInt();
      zbSwitch.setLightLevel(level);
    } else {
      Serial.println("Unknown command");
    }
  }
/*
  // print the bound devices (lights) every 30 seconds
  static uint32_t last_print = 0;
  if (millis() - last_print > 30000) {
    last_print = millis();
    zbSwitch.printBoundDevices(Serial);
  }

  */
}

**Comment by mwxp06** [comment]
Hi @P-R-O-C-H-Y  I would like to create some like weather station with ESP32c6 - it means multiple sensors (BME680, Light GY-30, Capacitive Soil Moisture..).
Can you give me an advice about some "Best Practices"? 

**Possible solutions:**
- create my own cpp files with combination all this sensors as single class and use it in ino file
- extend your library for missing sensor data e.g.: add Pressure, CO, into ZigbeeTempSensor (one class for one sensor)
- use existing library and create my own only for missing data..

But there is a big question if I can combinate multiple:
  // Add endpoint to Zigbee Core
    Zigbee.addEndpoint(&zbTempSensor);
    Zigbee.addEndpoint(&zbPressureSensor);
    Zigbee.addEndpoint(&zbLightSensor);
    ...
    
  If I am not wrong, you have not example where you are using multiple libraries in one file..
  
  Thank you very much for your help 
  and Marry christmas.

**Reply by IgorCode** [reply]
They have `Zigbee_Pressure_Flow_Sensor` which uses two endpoints `ZigbeeFlowSensor` and `ZigbeePressureSensor`.

And in case someone had doubts (like I did) for Home Assistant multiple endpoints get combined in a single device:
![image](https://github.com/user-attachments/assets/690806c5-cb95-4314-90c1-b3eceff994b9)
